// railway-passport/lib/strategies/foursquare.js
var passport = require('passport');

exports.callbackConnect = function(token, tokenSecret, profile, done) {
    console.info("staring callback for foursquare account connect auth")
    console.info(profile)
    console.info(token)
    console.info(tokenSecret)
    // User.findOne({where: { foursquareId: profile.id, foursquareAccessToken: token, foursquareAccessTokenSecret: token }}, function(err, account) {
    User.findOne({where: { foursquareId: profile.id, foursquareAccessToken: token }}, function(err, account) {
        if (err) { return done(err); }
        if (account) { return done(null, account) }

        var account = new User();
        account.foursquareId = profile.id;
        account.foursquareAccessToken = token;
        return done(null, account);
    });
};

exports.init = function (conf) {
    var Strategy = require('passport-foursquare').Strategy;
    passport.use(new Strategy({
        clientID: conf.foursquare.apiKey,
        clientSecret: conf.foursquare.secret,
        callbackURL: conf.baseURL + 'auth/foursquare/callback'
    }, exports.callback));

    passport.use('foursquare-connect', new Strategy({
        clientID: conf.foursquare.apiKey,
        clientSecret: conf.foursquare.secret,
        callbackURL: conf.baseURL + 'connect/foursquare/callback'
    }, exports.callbackConnect));

    // Redirect to Twitter for account authorization.
    app.get('/connect/foursquare',
      passport.authorize('foursquare-connect', { failureRedirect: '/connect/accounts' })
    );

    app.get('/connect/foursquare/callback',
      passport.authorize('foursquare-connect', { failureRedirect: '/connect/accounts' }),
      function(req, res) {
        var user = req.user;
        // var account = req.account;
        user.updateAttributes({
            foursquareId: req.account.foursquareId,
            foursquareAccessToken: req.account.foursquareAccessToken,
            updatedAt: new Date()
        },function(err){
            req.flash('info', 'foursquare account linked!');
            res.redirect('/connect/accounts');
        });
      }
    );

};
