// railway-passport/lib/strategies/goodreads.js
var passport = require('passport');

exports.callbackConnect = function(token, tokenSecret, profile, done) {
    console.info("staring callback for goodreads account connect auth")
    console.info(profile)
    console.info(token)
    console.info(tokenSecret)
    // User.findOne({where: { goodreadsId: profile.id, goodreadsAccessToken: token, goodreadsAccessTokenSecret: token }}, function(err, account) {
    User.findOne({where: { goodreadsId: profile.id, goodreadsAccessToken: token,  goodreadsAccessTokenSecret: tokenSecret }}, function(err, account) {
        if (err) { return done(err); }
        if (account) { return done(null, account) }

        var account = new User();
        account.goodreadsId = profile.id;
        account.goodreadsAccessToken = token;
        account.goodreadsAccessTokenSecret = tokenSecret;
        return done(null, account);
    });
};

exports.init = function (conf) {
    // console.log(conf.goodreads.apiKey)
    // console.log(conf.goodreads.secret)
    var Strategy = require('passport-goodreads').Strategy;
    // passport.use(new Strategy({
    //     consumerKey: conf.goodreads.apiKey,
    //     consumerSecret: conf.goodreads.secret,
    //     callbackURL: conf.baseURL + 'auth/goodreads/callback'
    // }, exports.callback));

    passport.use('goodreads-connect', new Strategy({
        consumerKey: conf.goodreads.apiKey,
        consumerSecret: conf.goodreads.secret,
        callbackURL: conf.baseURL + 'connect/goodreads/callback'
    }, exports.callbackConnect));

    // Redirect to Twitter for account authorization.
    app.get('/connect/goodreads',
      passport.authorize('goodreads-connect', { failureRedirect: '/connect/accounts' })
    );

    app.get('/connect/goodreads/callback',
      passport.authorize('goodreads-connect', { failureRedirect: '/connect/accounts' }),
      function(req, res) {
        var user = req.user;
        // var account = req.account;
        user.updateAttributes({
            goodreadsId: req.account.goodreadsId,
            goodreadsAccessToken: req.account.goodreadsAccessToken,
            goodreadsAccessTokenSecret: req.account.goodreadsAccessTokenSecret,
            updatedAt: new Date()
        },function(err){
            req.flash('info', 'goodreads account linked!');
            res.redirect('/connect/accounts');
        });
      }
    );

};
