// railway-passport/lib/strategies/linkedin.js
var passport = require('passport');

exports.callbackConnect = function(token, tokenSecret, profile, done) {
    console.info("staring callback for linkedin account connect auth")
    console.info(profile)
    console.info(token)
    console.info(tokenSecret)
    // User.findOne({where: { linkedinId: profile.id, linkedinAccessToken: token, linkedinAccessTokenSecret: token }}, function(err, account) {
    User.findOne({where: { linkedinId: profile.id, linkedinAccessToken: token,  linkedinAccessTokenSecret: tokenSecret }}, function(err, account) {
        if (err) { return done(err); }
        if (account) { return done(null, account) }

        var account = new User();
        account.linkedinId = profile.id;
        account.linkedinAccessToken = token;
        account.linkedinAccessTokenSecret = tokenSecret;
        return done(null, account);
    });
};

exports.init = function (conf) {
    // console.log(conf.linkedin.apiKey)
    // console.log(conf.linkedin.secret)
    var Strategy = require('passport-linkedin').Strategy;
    // passport.use(new Strategy({
    //     consumerKey: conf.linkedin.apiKey,
    //     consumerSecret: conf.linkedin.secret,
    //     callbackURL: conf.baseURL + 'auth/linkedin/callback'
    // }, exports.callback));

    passport.use('linkedin-connect', new Strategy({
        consumerKey: conf.linkedin.apiKey,
        consumerSecret: conf.linkedin.secret,
        callbackURL: conf.baseURL + 'connect/linkedin/callback'
    }, exports.callbackConnect));

    // Redirect to Twitter for account authorization.
    app.get('/connect/linkedin',
      passport.authorize('linkedin-connect', { failureRedirect: '/connect/accounts' })
    );

    app.get('/connect/linkedin/callback',
      passport.authorize('linkedin-connect', { failureRedirect: '/connect/accounts' }),
      function(req, res) {
        var user = req.user;
        // var account = req.account;
        user.updateAttributes({
            linkedinId: req.account.linkedinId,
            linkedinAccessToken: req.account.linkedinAccessToken,
            linkedinAccessTokenSecret: req.account.linkedinAccessTokenSecret,
            updatedAt: new Date()
        },function(err){
            req.flash('info', 'linkedin account linked!');
            res.redirect('/connect/accounts');
        });
      }
    );

};
