// railway-passport/lib/strategies/local.js
var passport = require('passport');

exports.callback = function (email, password, done) { 
    User.findOrCreate({
        email: email,
        password: password
    }, function (err, user) {
        console.log("in callback for find or create local strategy")
        if (err) {
            return done(err);
        }
        else
        if (!user) {
            // console.log('no account');
            return done(err, false,{message:"You don't have an account, please sign up"});
        }
        else
        // console.log(user) //form password , hashpassword
        if (!User.verifyPassword(password, user.password)) {
            // console.log('wrong password');
            return done(err, false,{message:'Your credentials do not match'});
        }
        else{
            // console.log('got it!');
            // console.log(user);
            return done(err, user);
        }
    });
};

exports.init = function (conf) {
    var Strategy = require('passport-local').Strategy;
    passport.use(new Strategy({
        usernameField: conf.usernameField || 'email'
    }, exports.callback));

    app.post('/login', passport.authenticate('local', {
        failureRedirect: '/login',
        failureFlash: true,
        badRequestMessage: 'Could not match credentials, please create an account or recover your password'
    }), exports.redirectOnSuccess);
};
