// railway-passport/lib/strategies/meetup.js
var passport = require('passport');

exports.callbackConnect = function(token, tokenSecret, profile, done) {
    console.info("staring callback for meetup account connect auth")
    console.info(profile)
    console.info(token)
    console.info(tokenSecret)
    // User.findOne({where: { meetupId: profile.id, meetupAccessToken: token, meetupAccessTokenSecret: token }}, function(err, account) {
    User.findOne({where: { meetupId: profile.id, meetupAccessToken: token,  meetupAccessTokenSecret: tokenSecret }}, function(err, account) {
        if (err) { return done(err); }
        if (account) { return done(null, account) }

        var account = new User();
        account.meetupId = profile.id;
        account.meetupAccessToken = token;
        account.meetupAccessTokenSecret = tokenSecret;
        return done(null, account);
    });
};

exports.init = function (conf) {
    // console.log(conf.meetup.apiKey)
    // console.log(conf.meetup.secret)
    var Strategy = require('passport-meetup').Strategy;
    // passport.use(new Strategy({
    //     consumerKey: conf.meetup.apiKey,
    //     consumerSecret: conf.meetup.secret,
    //     callbackURL: conf.baseURL + 'auth/meetup/callback'
    // }, exports.callback));

    passport.use('meetup-connect', new Strategy({
        consumerKey: conf.meetup.apiKey,
        consumerSecret: conf.meetup.secret,
        callbackURL: conf.baseURL + 'connect/meetup/callback'
    }, exports.callbackConnect));

    // Redirect to Twitter for account authorization.
    app.get('/connect/meetup',
      passport.authorize('meetup-connect', { failureRedirect: '/connect/accounts' })
    );

    app.get('/connect/meetup/callback',
      passport.authorize('meetup-connect', { failureRedirect: '/connect/accounts' }),
      function(req, res) {
        var user = req.user;
        // var account = req.account;
        user.updateAttributes({
            meetupId: req.account.meetupId,
            meetupAccessToken: req.account.meetupAccessToken,
            meetupAccessTokenSecret: req.account.meetupAccessTokenSecret,
            updatedAt: new Date()
        },function(err){
            req.flash('info', 'meetup account linked!');
            res.redirect('/connect/accounts');
        });
      }
    );

};
