// railway-passport/lib/strategies/soundcloud.js
var passport = require('passport');

exports.callbackConnect = function(token, tokenSecret, profile, done) {
    console.info("staring callback for soundcloud account connect auth")
    console.log(profile)
    User.findOne({where: { soundcloudId: profile.id, soundcloudAccessToken: token, soundcloudAccessTokenSecret: token }}, function(err, account) {
        if (err) { return done(err); }
        if (account) { return done(null, account) }

        var account = new User();
        account.soundcloudId = profile.id;
        account.soundcloudAccessToken = token;
        account.soundcloudAccessTokenSecret = tokenSecret;
        return done(null, account);
    });
};

exports.init = function (conf) {
    var Strategy = require('passport-soundcloud').Strategy;
    passport.use(new Strategy({
        clientID: conf.soundcloud.apiKey,
        clientSecret: conf.soundcloud.secret,
        callbackURL: conf.baseURL + 'auth/soundcloud/callback'
    }, exports.callback));

    passport.use('soundcloud-connect', new Strategy({
        clientID: conf.soundcloud.apiKey,
        clientSecret: conf.soundcloud.secret,
        callbackURL: conf.baseURL + 'connect/soundcloud/callback'
    }, exports.callbackConnect));

    // Redirect to Twitter for account authorization.
    app.get('/connect/soundcloud',
      passport.authorize('soundcloud-connect', { failureRedirect: '/connect/accounts', scope:"non-expiring" })
    );

    app.get('/connect/soundcloud/callback',
      passport.authorize('soundcloud-connect', { failureRedirect: '/connect/accounts', scope:"non-expiring" }),
      function(req, res) {
        var user = req.user;
        // var account = req.account;
        user.updateAttributes({
            soundcloudId: req.account.soundcloudId,
            soundcloudAccessToken: req.account.soundcloudAccessToken,
            soundcloudAccessTokenSecret: req.account.soundcloudAccessTokenSecret,
            updatedAt: new Date()
        },function(err){
            req.flash('info', 'Soundcloud account linked!');
            res.redirect('/connect/accounts');
        });
      }
    );

};
