// railway-passport/lib/strategies/tumblr.js
var passport = require('passport');

var TUMBLR_CONSUMER_KEY = "--insert-tumblr-consumer-key-here--"
var TUMBLR_SECRET_KEY = "--insert-tumblr-secret-key-here--";

exports.callbackConnect = function(token, tokenSecret, profile, done) {
    console.info("staring callback for tumblr account connect auth")
    // console.info(profile)
    // console.info(profile._json.response.user)
    // console.info(token)
    // console.info(tokenSecret)
    User.findOne({where: { 
            tumblrUsername: profile.username, 
            tumblrAccessToken: token, 
            tumblrAccessTokenSecret: tokenSecret }
        }, function(err, account) {
        if (err) { return done(err); }
        if (account) { return done(null, account) }

        var account = new User();
        account.tumblrUsername = profile.username;
        account.tumblrAccessToken = token;
        account.tumblrAccessTokenSecret = tokenSecret;
        return done(null, account);
    });
};

//kljhlk

exports.init = function (conf) {
    var Strategy = require('passport-tumblr').Strategy;
// console.info(conf.tumblr.apiKey)
// console.info(conf.tumblr.secret)

    passport.use('tumblr-connect', new Strategy({
        consumerKey: conf.tumblr.apiKey,
        consumerSecret: conf.tumblr.secret,
        callbackURL: conf.baseURL + 'connect/tumblr/callback'
    }, exports.callbackConnect));

    // Redirect to tumblr for account authorization.
    app.get('/connect/tumblr',
      passport.authorize('tumblr-connect', { failureRedirect: '/connect/accounts' })
    );

    app.get('/connect/tumblr/callback',
      passport.authorize('tumblr-connect', { failureRedirect: '/connect/accounts' }),
      function(req, res) {
        var user = req.user;
        var account = req.account;
        user.updateAttributes({
            tumblrUsername: req.account.tumblrUsername,
            tumblrAccessToken: req.account.tumblrAccessToken,
            tumblrAccessTokenSecret: req.account.tumblrAccessTokenSecret,
            updatedAt: new Date()
        },function(err){
            req.flash('info', 'tumblr account linked!');
            res.redirect('/connect/accounts');
        });
      }
    );

};
