// railway-passport/lib/strategies/tumblr.js
var passport = require('passport');

exports.callbackConnect = function(token, tokenSecret, profile, done) {
    console.info("staring callback for tumblr account connect auth")
    console.info(profile)
    console.info(token)
    console.info(tokenSecret)
    User.findOne({where: { 
            tumblrId: profile.id, 
            tumblrAccessToken: token }
        }, function(err, account) {
        if (err) { return done(err); }
        if (account) { return done(null, account) }

        var account = new User();
        account.tumblrId = profile.id;
        account.tumblrAccessToken = token;
        return done(null, account);
    });
};

exports.init = function (conf) {
    var Strategy = require('passport-tumblr').Strategy;
    passport.use(new Strategy({
        consumerKey: conf.tumblr.apiKey,
        consumerSecret: conf.tumblr.secret,
        callbackURL: conf.baseURL + 'auth/tumblr/callback'
    }, exports.callback));

    passport.use('tumblr-connect', new Strategy({
        consumerKey: conf.tumblr.apiKey,
        consumerSecret: conf.tumblr.secret,
        callbackURL: conf.baseURL + 'connect/tumblr/callback'
    }, exports.callbackConnect));

    // Redirect to tumblr for account authorization.
    app.get('/connect/tumblr',
      passport.authorize('tumblr-connect', { failureRedirect: '/connect/accounts' })
    );

    app.get('/connect/tumblr/callback',
      passport.authorize('tumblr-connect', { failureRedirect: '/connect/accounts' }),
      function(req, res) {
        var user = req.user;
        // var account = req.account;
        user.updateAttributes({
            tumblrId: req.account.tumblrId,
            tumblrAccessToken: req.account.tumblrAccessToken,
            // tumblrAccessTokenSecret: req.account.tumblrAccessTokenSecret,
            updatedAt: new Date()
        },function(err){
            req.flash('info', 'tumblr account linked!');
            res.redirect('/connect/accounts');
        });
      }
    );

};
